### ISSUE TITLE
How to be notified when player thread completes a user request

### ISSUE THREAD

**Issue by stevemayhew** [androidx/media#222]
Is there a way to only be notified to `Player.Listener` and/or `AnalyticsListener` when an operation is acknowledged, that is actually executed, by the main player thread (EPII)?

For example, if you issue a call to `Player.setPlaybackParameters()` there are two paths that can notify of the change:

1. From ExoPlayerImpl (Application Thread side) - [updatePlaybackInfo()](https://github.com/androidx/media/blob/c2cbb6370a265efc93661b659e70b2679506e995/libraries/exoplayer/src/main/java/androidx/media3/exoplayer/ExoPlayerImpl.java#L1818) 
2. From ExoPlayerImplInternal (Player Thread side) - [maybeNotifyPlaybackInfoChanged()](https://github.com/androidx/media/blob/c2cbb6370a265efc93661b659e70b2679506e995/libraries/exoplayer/src/main/java/androidx/media3/exoplayer/ExoPlayerImplInternal.java#L691)

Logging reveals what actually happens:

````
12-13 15:34:56.977 14003 14003 D ExoPlayerImpl: Player.setPlaybackParameters() - newParameters: PlaybackParameters(speed=15.00, pitch=1.00) oldParameters:PlaybackParameters(speed=1.00, pitch=1.00) looper: Looper (main, tid 2) {aeca7b7}
12-13 15:34:56.977 14003 14003 D ExoPlayerImplInternal: setPlaybackParameters() - parameters: PlaybackParameters(speed=15.00, pitch=1.00) looper: Looper (ExoPlayer:Playback, tid 53090) {675b24}
12-13 15:34:56.977 14003 14003 D ExoPlayerImpl: updatePlaybackInfo() - parameters changed, old: PlaybackParameters(speed=1.00, pitch=1.00) new: PlaybackParameters(speed=15.00, pitch=1.00)
12-13 15:34:56.978 14003 14003 D StateTransitionManager: onPlaybackParametersChanged() - PlaybackParameters(speed=15.00, pitch=1.00)
12-13 15:34:57.069 14003 14132 D ExoPlayerImplInternal: setPlaybackParametersInternal() - playbackParameters: PlaybackParameters(speed=15.00, pitch=1.00)
12-13 15:34:57.071 14003 14132 D ExoPlayerImplInternal: maybeNotifyPlaybackInfoChanged() - hasPendingChange, prev: PlayackInfo@71aff36 - parameters: PlaybackParameters(speed=1.00, pitch=1.00) state: 3 new: PlayackInfo@a054e37 - parameters: PlaybackParameters(speed=15.00, pitch=1.00) state: 3
````

As you can see, the notification delivered by EPI from the same call stack that called `Player.setPlaybackParameters()`, not reentrant because of the `ListenerSet` changes (very helpful).  But, it is nearly 100ms before the actual speed change is committed to the player.  The listener method called in `StateTransitionManager` is an `AnalyticsListener` method (`Player.Listener` gives the same result).

The last log line shows the change in EPII's version of `PlaybackInfo`, here the change is posted in a runnable back to the Application Looper to the method [handlePlaybackInfo()](https://github.com/androidx/media/blob/c2cbb6370a265efc93661b659e70b2679506e995/libraries/exoplayer/src/main/java/androidx/media3/exoplayer/ExoPlayerImpl.java#L1762) however by then the `PlaybackInfo` in EPI is already updated so no change event is delivered.

So basically:

1. There is a window where EPI PlaybackParameters are different from EPII (this is of course un-avoidable)
2. It is not possible to know when that window closes (easily at least)

The only way I can see without changing this code is to use `SimpleExoPlayer.createMessage()`, but that seems like a hack workaround.

Or we could accomplish 2 by some experimental parameter to turn of the first notification (call to  [updatePlaybackInfo()](https://github.com/androidx/media/blob/c2cbb6370a265efc93661b659e70b2679506e995/libraries/exoplayer/src/main/java/androidx/media3/exoplayer/ExoPlayerImpl.java#L1818)).

Why this matters for us is a long discussion, but it centers around how Live Edge Adjustment manipulates playback speed.  @tonihei  your thoughts?

**Comment by tonihei** [comment]
> But, it is nearly 100ms before the actual speed change is committed to the player.

This is actually very surprising and not generally expected. The difference between the two log messages `ExoPlayerImplInternal: setPlaybackParameters()` and `ExoPlayerImplInternal: setPlaybackParametersInternal()` should just be a simple `Handler` message passed to another thread and I expect that shouldn't take longer than a few milliseconds. The only reason I see why this isn't the case is that the playback thread was blocked on some operation that took almost 100 ms. The main example that comes to mind is starting a decoder. Do you call this method around the same time as playback starts and the playback thread might be busy setting up a decoder? 

It would be useful to understand why you want to know when this happens because the API is designed in a way that you shouldn't have to care. Basically, from an app's point of view, you set the speed and the player updates the state and calls its listeners immediately. The fact that there might be a small delay before any command takes effect on another thread is usually not relevant.

> The only way I can see without changing this code is to use SimpleExoPlayer.createMessage(), but that seems like a hack workaround.

This is actually a sensible way of doing this, as long as you set the message looper to be the main looper. This way the callback of the message happens after all previously issued actions are triggered internally and are completely handled. We actually use the [same technique](https://github.com/google/ExoPlayer/blob/0c74758f0e6d5acff523182ac93ac7b91324498d/robolectricutils/src/main/java/com/google/android/exoplayer2/robolectric/TestPlayerRunHelper.java#L317) in various unit tests where we want to assert that nothing happens (`do action A`, `sendMessage`, `waitForMessage`, `assertThatADidNotDoSomething`)

**Comment by stevemayhew** [comment]
Thanks, I'll use the `createMessage()` technique then.  IMO, the path for change notification via [updatePlaybackInfo()](https://github.com/androidx/media/blob/c2cbb6370a265efc93661b659e70b2679506e995/libraries/exoplayer/src/main/java/androidx/media3/exoplayer/ExoPlayerImpl.java#L1818) isn't needed, if it was not done then the event would come from the EPII side.

**Comment by stevemayhew** [comment]
Thanks for the help!
